### PR TITLE
Docs pass

### PR DESCRIPTION
**Pull request by dgasmith**
## Description
Starts a high-level organization of the QCFractal documentation:
 - Glossary
 - Portal client
 - Portal molecule 

Several other issues:
 - Fixed an error where non-ObjectID index queries would result in a server raise.
 - Fixed an error where molecule with fragments, but without fragment mult/charge info would cause a code error. These are now correctly raised and the canonical case is automatically generated.

## Status
- [x] Ready to go

### CODE REVIEW

**Comment by dgasmith** [comment]
My last doc changes for a bit, this is ready for a review.

File: docs/source/fractal-procedures.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Procedures
```
Comment (loriab) [inline docs/source/fractal-procedures.rst]: Note that the file names become the html names (e.g., `fractal-procedures.html`) and ppl link to those. Fine by me, just a heads-up if you want fractal everywhere.

File: docs/source/fractal-results.rst [hunk 2]
```diff
@@ -0,0 +1,94 @@
+Results
+========
+
+Results are a single quantum chemistry method evaluation including energy,
+gradient, Hessian, and property evaluations. Collections of evaluations such
+as finite-difference gradients, complete basis set extrapolation, or geometry
+optimizations belong under the "Procedures" heading.
```
Comment (loriab) [inline docs/source/fractal-results.rst]: "A result is a single quantum chemistry method evaluation, which might be an energy, an analytic gradient or Hessian, or a property evaluation."

File: docs/source/fractal-results.rst [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+Schema
+------
+
+All results are stored using the `QCSchema <https://molssi-qc-schema.readthedocs.io/en/latest/index.html>`_ so that the storage is quantum chemistry program agnostic. An example of the QCSchema input is shown below:
+
+.. code-block:: python
+
+    {
+      "schema_name": "qc_json_input",
+      "schema_version": 0,
```
Comment (loriab) [inline docs/source/fractal-results.rst]: 0-->1

File: docs/source/fractal-services.rst [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+ 1. A client submits a new service request to the server.
+ 2. A service is created on the server and place in the service queue.
+ 3. A service iteration is called that will spawn new tasks.
+ 4. The service waits until all generated tasks are complete.
+ 5. The service repeats 3 and 4 until the service iterations are complete.
+ 6. The service cleans intermediate data, finalizes the data representation, and marks itself complete.
+
+To give an example, the TorsionDrive Service will be used. This service optimizes the geometry of a biomolecule at a number
+of frozen dihedral angles to provide an energy profile of the rotation of this dihedral bond. Here the geometry
```
Comment (loriab) [inline docs/source/fractal-services.rst]: "The TorsionDriver Service is an example. The service"

Dangling "Here the geometry"

File: docs/source/fractal-services.rst [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+ 1. A client submits a task to scan the HOOH molecule dihedral every 90 degrees.
+ 2. The service is received by the server, and the first 0-degree dihedral geometry optimization is spawned.
+ 3. The service waits until the 0-degree task is complete, and then generate 90 and -90-degree tasks based off this 0-degree geometry.
+ 4. The service waits for the two new jobs to complete and spawns 0 and 180-degree tasks based on the 90 and - 90-degree geometries.
+ 5. The 90 and -90-degree tasks complete, the service builds its final data structure for user querying and marks itself complete.
```
Comment (loriab) [inline docs/source/fractal-services.rst]: Plural "generates" in item 3.

"5. The service waits for the 90- and -90-degree tasks to complete. Then it builds its final ..."

File: docs/source/glossary.rst [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+      On-node computations, these can either be a single computation (energy, gradient, property, etc.) or a series of calculations such as a geometry optimization.
+
+    Services
+      Iterative workflows where the required computations are distributed via the queue and then are processed on the server to acquire the next iteration of calculations.
+
+    Queue Handlers
+      The interface between QCFractal's internal queue representation and other queueing systems such as Dask or Fireworks.
+
+    Molecule
+      A unique 3D representation of a molecule. Any changes to the protonation state, multiplicity, charge, fragments, coordinates, connectivity, or
```
Comment (loriab) [inline docs/source/glossary.rst]: isotope

File: docs/source/index.rst [hunk 7]
```diff
@@ -55,20 +55,64 @@ The QCArchive project's primary support comes from `The Molecular Sciences Softw
...
+   :hidden:
    :caption: Getting Started
 
    install
    community
    roadmap
 
+**Portal Documentation**
+
+Portal is the primary user interface to a Fractal server instance.
```
Comment (loriab) [inline docs/source/index.rst]: I recc. defining a `|qcf|` and `|qcfs|` for "QCFractal" and "QCFractal's" in conf.py and using it everywhere. Shorter to type (unlike `|PSIfour|`) and ensures uniformity throughout docs.

File: docs/source/portal-client.rst [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    >>> import qcfractal.interface as portal
+    >>> client = portal.FractalClient("localhost:8888")
+    >>> client
+    FractalClient(server='http://localhost:8888/', username='None')
+
+The ``FractalClient`` handles all communication to the server from the Python API layer. This core communication access the fundamental structure of the server is often not used in favor of more general handlers. See (Database/TorsionDriveORM/etc).
+
+The ``FractalClient`` can also be initialized from a file which is useful so that addresses and username do not have to be retyped for every line and reduces the chance that a username and password could accidentally be added to a version control system. This can be accessed by creating a new object by ``FractalClient.from_file()`` whereby default the server searches for a ``qcportal_config.yaml`` file in either the current working directory or from the canonical ``~/.qca/`` folder.
+
+Molecule Handling
```
Comment (loriab) [inline docs/source/portal-client.rst]: you'll want the below over every section of interest, so you can refer to it elsewhere in docs via `:ref:`

```
.. _`sec:introduction`:
```

File: docs/source/portal-client.rst [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+the setting up a server (NYI) section.
+
+.. code-block:: python
+
+    >>> import qcfractal.interface as portal
+    >>> client = portal.FractalClient("localhost:8888")
+    >>> client
+    FractalClient(server='http://localhost:8888/', username='None')
+
+The ``FractalClient`` handles all communication to the server from the Python API layer. This core communication access the fundamental structure of the server is often not used in favor of more general handlers. See (Database/TorsionDriveORM/etc).
```
Comment (loriab) [inline docs/source/portal-client.rst]: Does the second sentence make sense?

File: docs/source/portal-client.rst [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+The ``FractalClient`` can also be initialized from a file which is useful so that addresses and username do not have to be retyped for every line and reduces the chance that a username and password could accidentally be added to a version control system. This can be accessed by creating a new object by ``FractalClient.from_file()`` whereby default the server searches for a ``qcportal_config.yaml`` file in either the current working directory or from the canonical ``~/.qca/`` folder.
```
Comment (loriab) [inline docs/source/portal-client.rst]: "In ``FractalClient.from_file()``, by default the server searches for a ``qcportal_config.yaml`` file in either the current working directory or from the canonical ``~/.qca/`` folder."

File: docs/source/portal-client.rst [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    Adding the same molecule to the database as an idempotent operation and returns the same
+    molecule ID.
+
+    .. code-block:: python
+
+        >>> data = client.add_molecules({"hooh": hooh})
+        >>> data
+        {'hooh': '5b882c957b87878925ffaf22'}
+
+Molecules can either be queried based of their Molecule ID or their Molecule hash:
```
Comment (loriab) [inline docs/source/portal-client.rst]: "queried from Molecule ID or Molecule hash:"

File: docs/source/portal-molecule.rst [hunk 12]
```diff
@@ -0,0 +1,84 @@
+Portal Molecule
+===============
+
+The Portal Molecule class can represent many different quantum chemical molecules and also provides the primary manipulation tool for molecules.
+
+Creating Molecules
+------------------
+
+Molecules can be created via NumPy, Psi4-like molecule strings, and QCSchema JSON representations.
```
Comment (loriab) [inline docs/source/portal-molecule.rst]: "QCJson_Schema" is another thing that needs a `|qcjson|` alias. I can see its official name changing.

File: docs/source/portal-molecule.rst [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        Geometry (in Angstrom), charge = 0.0, multiplicity = 1:
+
+           Center              X                  Y                   Z
+        ------------   -----------------  -----------------  -----------------
+              He          0.000000000000     0.000000000000    -1.587531625770
+              He          0.000000000000     0.000000000000     1.587531625770
+
+.. note::
+
+    Molecules can automatically be orientated to a common frame of reference using the ``orient=True`` kwarg.
```
Comment (loriab) [inline docs/source/portal-molecule.rst]: need to bring this in line with qcjson (`fix_orientation`)

File: docs/source/portal-molecule.rst [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+    >>> mol.get_hash()
+    '84872f975d19aafa62b188b40fbadaf26a3b1f84'
+
+.. note::
+
+    A molecule hash is different from a Molecule ID as the Molecule ID is a
+    :term:`DB Index` and reference where the molecule is in the database and
+    is different between databases. A molecule hash is a unique molecule
+    identifier that can be created and accessed anywhere.
```
Comment (loriab) [inline docs/source/portal-molecule.rst]: "from a Molecule ID, as the latter is a :term:`DB Index` referencing membership in the database (and thus is different between databases). A molecule hash ..."

File: docs/source/portal-molecule.rst [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+
+Fragments
+---------
+
+The molecule class natively supports fragments. If a molecule is built with fragments, obtained an individual piece is straightforward.
```
Comment (loriab) [inline docs/source/portal-molecule.rst]: obtaining

File: docs/source/portal-molecule.rst [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+.. code-block:: python
+
+    dimer = portal.Molecule({"symbols": ["He", "He"], "geometry": [0, 0, -3, 0, 0, 3], "fragments": [[0], [1]]})
+
+    >>> dimer.get_fragment(0)
+        Geometry (in Angstrom), charge = 0.0, multiplicity = 0:
```
Comment (loriab) [inline docs/source/portal-molecule.rst]: multiplicity of 0 is suspicious

File: qcfractal/interface/client.py [hunk 17]
```diff
@@ -31,6 +45,19 @@ def __init__(self, address, username=None, password=None, shared_secret=None):
         else:
             self._auth = {}
 
+    def __repr__(self):
```
Comment (loriab) [inline qcfractal/interface/client.py]: Should be `__str__` instead? `__repr__` often is a string that can be used to re-instantiate the object.

File: qcfractal/interface/molecule.py [hunk 18]
```diff
@@ -26,7 +26,22 @@ class Molecule:
...
+        ----------
+        mol_str : str, dict
+            A string or dictionary representation of the molecule.
+        name : {"", str}, optional
+            The name of the molecule, defaults to "".
+        dtype : {None, "numpy", "json", "psi4"}, optional
+            Determines how the input is parsed.
+        orient : bool, optional
+            Orientate the molecule to a standard frame or not.
+
```
Comment (loriab) [inline qcfractal/interface/molecule.py]: Can we consolidate with molparse someday?

**Comment by loriab** [comment]
Oh, and it's worth running `<F2>` or whatever on the docs paragraphs to break them into lines for `git diff`'s sake in future.

**Comment by andysim** [comment]
In vi it's, e.g., `2gqq` to break up two lines and  `3gqq` for three lines, etc.

File: docs/source/fractal-procedures.rst (hunk 1 above)
Comment (dgasmith) [reply docs/source/fractal-procedures.rst]: I go back and forth on this topic, not really sure which one I like better. Portal and Fractal are divided on these lines for experimentation.

File: docs/source/portal-client.rst (hunk 9 above)
Comment (dgasmith) [reply docs/source/portal-client.rst]: Does not :)

File: qcfractal/interface/molecule.py (hunk 18 above)
Comment (dgasmith) [reply qcfractal/interface/molecule.py]: Very much so!

**Pull request merged**